ZooKeeper module: Adapt to ZooKeeper 3.6+ `mntr` response fields' changes.

pfcoperez opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

This PR tries to tackle the compatibility issues that arise from trying to use the ZooKeeper metrics beats module with recent versions of this service: 3.6 and 3.7 majors.

Since ZooKeeper 3.6, some response to `mntr` 4lw commands requests have changed their implicit type from `integers` to `floats`. Fields like `zk_avg_latency`, `zk_min_latency` and `zk_max_latency` are some examples of this.

Additionally, `zk_followers` has been replaced by `zk_learners`.

This PR adds compatibility with these changes while not breaking pre ZK 3.6 ones.

## Why is it important?

It is key for Elastic cloud operations to have these changes in beats. This can surely be extrapolated to other users which might wonder why certain fields are not longer ingested after having upgraded to any of the most recent ZooKeeper majors.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved 

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

```bash
cd metricbeat
MODULE=zookeeper mage -v pythonIntegTest
```

## Related issues

Closes: https://github.com/elastic/beats/issues/30066

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- 

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
## :broken_heart: Tests Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30068/detail/PR-30068/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30068/detail/PR-30068/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30068/detail/PR-30068/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30068/detail/PR-30068/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_30068.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-01-27T17:24:38.086Z&rangeTo=2022-01-27T17:44:38.086Z&transactionName=Beats/beats/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=9ea1c149aad5ce5ea97ec5cc356259f5&transactionId=16757ef0d8db70fb)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats



* Reason: `null`


* Start Time: 2022-01-27T17:34:38.086+0000


* Duration: 45 min 13 sec


* Commit: c8ddbec04ebca4fb06d41fac5fb53c7b30a18e6c



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 1  |
| Passed       | 2458  |
| Skipped      | 566 |
| Total        | 3025   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30068/detail/PR-30068/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `Build&Test / metricbeat-goIntegTest / TestFetch – github.com/elastic/beats/v7/metricbeat/module/zookeeper/mntr`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestFetch
    mntr_integration_test.go:43: zookeeper/mntr event: {
          "service": {
            "node": {
              "name": "0"
            },
            "type": "zookeeper",
            "version": "3.5.5-390fe37ea45dee01bf87dc1c042b5e3dcce88653, built on 05/03/2019 12:07 GMT"
          },
          "zookeeper": {
            "mntr": {
              "approximate_data_size": 44,
              "ephemerals_count": 0,
              "latency": {
                "avg": 0,
                "max": 0,
                "min": 0
              },
              "max_file_descriptor_count": 1048576,
              "num_alive_connections": 1,
              "open_file_descriptor_count": 59,
              "outstanding_requests": 0,
              "packets": {
                "received": 12,
                "sent": 11
              },
              "server_state": "standalone",
              "watch_count": 0,
              "znode_count": 5
            }
          }
        }
--- FAIL: TestFetch (0.74s)
panic: interface conversion: interface {} is float64, not int64 [recovered]
	panic: interface conversion: interface {} is float64, not int64

goroutine 12 [running]:
testing.tRunner.func1.2({0xe59b80, 0xc00059c210})
	/usr/local/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1212 +0x218
panic({0xe59b80, 0xc00059c210})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/elastic/beats/v7/metricbeat/module/zookeeper/mntr.TestFetch(0xc000102ea0)
	/go/src/github.com/elastic/beats/metricbeat/module/zookeeper/mntr/mntr_integration_test.go:49 +0x795
testing.tRunner(0xc000102ea0, 0xf78f38)
	/usr/local/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1306 +0x35a
 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->











### Steps errors [![5](https://img.shields.io/badge/5%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30068/detail/PR-30068/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `metricbeat-unitTest - mage build unitTest`
<ul>
<li>Took 3 min 28 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-30068/runs/1/steps/3299/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `metricbeat-goIntegTest - mage goIntegTest`
<ul>
<li>Took 3 min 20 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-30068/runs/1/steps/3329/log/?start=0">here</a></li>
<li>Description: <code>mage goIntegTest</code></l1>
</ul>
  
  
  
  
##### `metricbeat-goIntegTest - mage goIntegTest`
<ul>
<li>Took 1 min 34 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-30068/runs/1/steps/3498/log/?start=0">here</a></li>
<li>Description: <code>mage goIntegTest</code></l1>
</ul>
  
  
  
  
##### `metricbeat-goIntegTest - mage goIntegTest`
<ul>
<li>Took 1 min 34 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-30068/runs/1/steps/3521/log/?start=0">here</a></li>
<li>Description: <code>mage goIntegTest</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-30068/runs/1/steps/3526/log/?start=0">here</a></li>
<li>Description: <code>Error 'hudson.AbortException: script returned exit code 1'</code></l1>
</ul>
  
  </p>
  </details>


## :bug: Flaky test report
:grey_exclamation: There are test failures but not known flaky tests.



<details><summary>Expand to view the summary</summary>
<p>

### Genuine test errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30068/detail/PR-30068/1//tests)
:broken_heart: There are test failures but not known flaky tests, most likely a genuine test failure.

* **Name**: `Build&Test / metricbeat-goIntegTest / TestFetch – github.com/elastic/beats/v7/metricbeat/module/zookeeper/mntr`

</p>
</details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/siem (Team:SIEM)

elasticmachine commented:
Pinging @elastic/integrations (Team:Integrations)

ruflin commented:
I assume this will need backport to 8.0 and 7.17, not 7.12?

mtojek commented on metricbeat/module/zookeeper/mntr/data.go:
```diff
@@ -36,9 +36,9 @@ var (
-			"avg": c.Int("zk_avg_latency"),
-			"min": c.Int("zk_min_latency"),
-			"max": c.Int("zk_max_latency"),
+			"avg": c.Float("zk_avg_latency"),
```
@ruflin This is a breaking change for zookeeper integration, as [integration](https://github.com/elastic/integrations/blob/main/packages/zookeeper/data_stream/mntr/fields/fields.yml) expects `long`. I don't think it's safe to push it.

sayden replied:
Yeah, I wanted to look at this if it's a permanent bug or a version mismatch

sayden replied:
Oh yes, it's a version mismatch from zookeeper 3.5 to 3.7. We'll need to upgrade the module and add the new version to the testing image in the CI

jsoriano replied:
Remember to update the `fields.yml` file too.

If this change is breaking, we should duplicate the field instead of modifying it.

sayden replied:
We cannot "just" duplicate the field because those keys must be unique. So instead it should be something like `avg_f`. Does that work for you @pfcoperez ?

The problem is the mapping, because we are producing a JSON output and JSON specification has no notion of integer numeric values anyways.

@jsoriano @mtojek have you ever tried to add 2 mapping types for the same metricbeat field? I know we can do it at elasticsearch level because we can set a field like `keyword` and `text` at the same time.

jsoriano replied:
> We cannot "just" duplicate the field because those keys must be unique.

Yes, this is what I meant, sorry :slightly_smiling_face: 

> have you ever tried to add 2 mapping types for the same metricbeat field?

I don't think this is possible, a field in an elasticsearch index can have only one mapping.

> I know we can do it at elasticsearch level because we can set a field like `keyword` and `text` at the same time.

I guess you refer to "[multi-fields](https://www.elastic.co/guide/en/elasticsearch/reference/7.16/multi-fields.html)". We could try to add a `float` field inside the current `long`, but this may be a bit hacky, the purpouse of this feature is different to what we need here. To access the float value you would access something like `...avg.float`.

sayden replied:
Oh yeah, I was refering to multi-fields indeed. yeah so we are in the same crossroad.

So let's wait to see @pfcoperez response to what works better for them in terms of field names and I'll update the PR myself.

pfcoperez replied:
I think we can keep the field as float. `0` will be parsed as `0.0`. In fact `ZK 3.7` sometimes sends ingeger formated values. e.g: `3`.

So, in a nutshell, if we changed these fields to `Float` we would parse ZK 3.6+ and previous responses alike, right?

ruflin replied:
Trying to understand on my end why the change from Int to Float is a breaking change?

pfcoperez replied:
Thanks for checking @ruflin , the breaking change happens at ZooKeeper side in relation to the beats module.
Currently, this module tries and fails to parse these metrics as integers. Until 3.5, zookeeper provided only integer values for the latency metrics.

However, since 3.7, it occasionally (whole values are not represented as `X.0`) responds with float number strings.

Take this example:

- ZK 3.5:

```
zk_avg_latency	0
zk_max_latency	29
zk_min_latency	0
```

- ZK 3.7:

```
zk_avg_latency	6.1223
zk_min_latency	0
zk_max_latency	73
```

I've checked and it seems that the beats parser works correctly when it tries to parse a whole number string (e.g `42` ) as a `Float`: `c.Float("zk_avg_latency"),` of course, it also succeeds in parsing float number strings (e.g `42.3`) so changing these fields to `Float` will make the module compatible with all ZooKeeper versions.

jsoriano replied:
> Trying to understand on my end why the change from Int to Float is a breaking change?

@ruflin it could be breaking on ingestion because it is a change on field types (from long to float), it could be also breaking if something as a dashboard is consuming these values as long, and they start being stored as floats. But I think we are safe here as floats can be ingested into long fields. If this value is used in dashboards, it will probably also work with floats. 

We can also play safer, and don't do the change in the mapping, then these values will be collected, but only its integer part would be ingested.

To store them as the floats they are, we need to change the data types in `metricbeat/module/zookeeper/mntr/_meta/fields.yml` from long to float.

The question would be if we only want this to "work", then the change to `Float` here is enough and it probably doesn't produce breaking changes, or if we want to also index it as proper `float`, then we need to change the mapping, and we need to test that nothing breaks. If something breaks, we need to do something as storing the new value in a new field.

pfcoperez replied:
> How valuable are the digits after the comma?

For Elastic cloud they are not super valuable, we can live without them. However having them is nice because visualizations quality will improve. For example, a `0.0` latency is extremely improbable and it might be interesting to discriminate between that situation and `0.XY` (more probable cases).

Anyway, So far we have lived  with just the integer part (as ZK didn't went further) and that's OK. A different thing is: How valuable are these fractional digits to other users? I don't know.

> We can also play safer, and don't do the change in the mapping, then these values will be collected, but only its integer part would be ingested.

With the above mentioned, if this is safe, this can be a nice sweet spot to fix Cloud monitoring issues and if any of our users need the decimal digits we can always follow up in future releases.

jsoriano replied:
Ok, then I think that we can go on with the changes here, so we at least support the newer versions, and we can consider in a follow up to properly index the floats.

sayden commented:
7.12 too so Cloud can do an progressive upgrade

sayden commented:
All right I'll have to generate a switch in the code flow depending on the zookeeper version

jsoriano reviewed (commented):
> 7.12 too so Cloud can do an progressive upgrade

If backported to 7.12, it should be backported first to all the branches in between (7.13, 7.14...), to avoid someone using the fixed 7.12 version finding the problem again when upgrading to one of the others.

In any case remember that very probably there is not going to be any other 7.12 release, so this would only fix the issue for custom builds.

ruflin commented:
There will not be any future 7.12 release. Even if we backport it, it never becomes available as a release. To get the change, Cloud has to go to 7.16, likely 7.17. @pfcoperez Is this an issue?

pfcoperez commented:
> There will not be any future 7.12 release. Even if we backport it, it never becomes available as a release. To get the change, Cloud has to go to 7.16, likely 7.17. @pfcoperez Is this an issue?

I think this might be an issue for ECE, would it be possible to do a minor official release?

pfcoperez commented:
I'd like to include a run of the ITs with ZK 3.7 (in addition to the current 3.5). Would you mind pointing me into the steps to do it? :bow:

ruflin commented:
Here is the compose file that spins up the environment: https://github.com/elastic/beats/blob/master/metricbeat/module/zookeeper/docker-compose.yml and the docker file: https://github.com/elastic/beats/blob/master/metricbeat/module/zookeeper/_meta/Dockerfile Interesting enough, it already has a variable for the version. @jsoriano I remember some services we tests multiple versions, do you have an example as reference?

jsoriano commented:
> I remember some services we tests multiple versions, do you have an example as reference?

Yep, we have support in the python tests. Take a look for example to the mysql or redis modules. You need these things:
* Add the `@metricbeat.parameterized_with_supported_versions` decorator to the test class in `test_zookeeper.py`.
* Parameterize the version or other variants in the docker compose file for zookeeper. I think that this is already done, but consider reviewing default values there, as they are the ones used by Go tests.
* Add a `_meta/supported-versions.yml` file, with each one of the versions to test, take a look to the [redis](https://github.com/elastic/beats/blob/dd24764c4a9ff57b52d0289f9647fee8a5c9358e/metricbeat/module/redis/_meta/supported-versions.yml) or [mysql](https://github.com/elastic/beats/blob/dd24764c4a9ff57b52d0289f9647fee8a5c9358e/metricbeat/module/mysql/_meta/supported-versions.yml) ones as example.

You can then run the python tests for this module with `MODULE=zookeeper mage -v pythonIntegTest`.

ruflin commented:
@sayden As a follow up, we should also update the compatibility of the zookeeper package: https://github.com/elastic/integrations/tree/main/packages/zookeeper

pfcoperez commented:
> > I remember some services we tests multiple versions, do you have an example as reference?
> 
> Yep, we have support in the python tests. Take a look for example to the mysql or redis modules. You need these things:
> 
>     * Add the `@metricbeat.parameterized_with_supported_versions` decorator to the test class in `test_zookeeper.py`.
> 
>     * Parameterize the version or other variants in the docker compose file for zookeeper. I think that this is already done, but consider reviewing default values there, as they are the ones used by Go tests.
> 
>     * Add a `_meta/supported-versions.yml` file, with each one of the versions to test, take a look to the [redis](https://github.com/elastic/beats/blob/dd24764c4a9ff57b52d0289f9647fee8a5c9358e/metricbeat/module/redis/_meta/supported-versions.yml) or [mysql](https://github.com/elastic/beats/blob/dd24764c4a9ff57b52d0289f9647fee8a5c9358e/metricbeat/module/mysql/_meta/supported-versions.yml) ones as example.
> 
> 
> You can then run the python tests for this module with `MODULE=zookeeper mage -v pythonIntegTest`.

Thank you @jsoriano ! I'll try to do it. I'll get back to you if I don't manage to.

pfcoperez commented:
@jsoriano Thanks to your indications, in 8c9c62cbbf1f39d766b68349454d31ffb6d831eb, I've added the validation against ZooKeeper 3.6 and 3.7, in addition to the current 3.5.

![image](https://user-images.githubusercontent.com/273379/152377457-81d75709-eef7-4958-a0a1-d83a3cbb4c3a.png)

:memo: Note that I've changed the source image for ZooKeeper to the official ones: https://hub.docker.com/_/zookeeper Is that OK?

jsoriano commented on metricbeat/module/zookeeper/_meta/supported-versions.yml:
```diff
@@ -0,0 +1,7 @@
+variants:
+  - ZOOKEEPER_VARIANT: zookeeper
```
`ZOOKEEPER_VARIANT` not needed?

pfcoperez replied:
Actually not, I thought that field was mandatory but I've tested removing it and it still works 2060c082ac712c3d579a7a9234eb0acd81583a1d

jsoriano commented on metricbeat/module/zookeeper/_meta/Dockerfile:
```diff
@@ -1,3 +1,3 @@
 ARG ZOOKEEPER_VERSION
-FROM jplock/zookeeper:${ZOOKEEPER_VERSION}
+FROM zookeeper:${ZOOKEEPER_VERSION}
```
:+1:

jsoriano commented on metricbeat/module/zookeeper/mntr/data.go:
```diff
@@ -53,7 +53,8 @@ var (
 	}
 	schemaLeader = s.Schema{
-		"followers":        c.Int("zk_followers"),
+		"learners":         c.Int("zk_learners", s.Optional),
```
This new `learners` field needs to be defined in `metricbeat/module/zookeeper/mntr/_meta/fields.yml`.

jsoriano replied:
> I've edited the description of the legacy field _followers_, is that OK?

Perfect, yes.

CI is failing because some files need to be updated after the change in fields.
Please run `make update` from the `metricbeat` and `x-pack/metricbeat` directories and commit the changes.

pfcoperez replied:
It is weir because it still  fails after doing so, going to force a new build.

jsoriano reviewed (commented):
Thanks! This is looking good. If tests pass I think that the only important pending things would be to:
* Add the new `learners` field to the `fields.yml` file. 
* Add a changelog entry.
* Update the compatibility notes in `metricbeat/module/zookeeper/_meta/docs.asciidoc`.

pfcoperez commented:
Thanks for the pointers @jsoriano , I've addressed the following issues:

>     * Add the new `learners` field to the `fields.yml` file.

ff855f3e64ece4e458b0a4c074c31a2df7f468a8

>     * Add a changelog entry.

b333d506e253985e5ddc18bc9c80b45edfda5126
 
>     * Update the compatibility notes in `metricbeat/module/zookeeper/_meta/docs.asciidoc`.

1c422416b17f71bc48bb29ca05daf9e2d32a7be5

jsoriano commented on metricbeat/module/zookeeper/mntr/data.go:
```diff
@@ -53,7 +53,8 @@ var (
 	schemaLeader = s.Schema{
-		"followers":        c.Int("zk_followers"),
+		"learners":         c.Int("zk_learners", s.Optional),
+		"followers":        c.Int("zk_followers", s.Optional), // Not present anymore in ZooKeeper >= 3.6 mntr responses
```
Umm, if this is a field that has been renamed, from followers to learners, we can keep both fields, and eventually remove the old one.

There was a similar case in redis, where some fields were renamed in 5.0. On #8167 we [renamed](https://github.com/elastic/beats/pull/8167/files#diff-55387d78bdbcd48ce8f65e10c3c2d4bce39e8c406b33212ccff10b9caecdf35eR83) the fields, storing the values obtained from each one of the versions on both fields. Eventually we removed the old field in https://github.com/elastic/beats/pull/28246, keeping compatibility with older versions by copying the old metric into the new field.

The advantage of this is that dashboards or other queries can work the same, independently of the Zookeeper version.

Wdyt about following a similar approach here?

(This could be done in a follow-up in any case)

pfcoperez replied:
Yes, exactly that case! So just to make sure I understood you proposal: The idea is to add a new "ingested" field which represents both source fields?

The current approach in this PR is: Add the new source field as a new ingested field, keep them both for compatibility purposes.

We would need to change the dashboards either way.

Or is it maybe that we would have two ingested fields "followers" (old) and "learners" (new) fed with the same whatever source field is present?

jsoriano replied:
The idea would be to expect the same fields in the documents, abstracting the Zookeeper version, so dashboard or queries work the same.

> Or is it maybe that we would have two ingested fields "followers" (old) and "learners" (new) fed with the same whatever source field is present?

Yes, it would be like that, to have the two fields always ingested. On old versions with the value of "zk_followers", and in new versions with the value of "zk_learners". We would deprecate "followers" and eventually remove it, but we could continue filling "learners" with the value of "zk_followers" in old versions.

Another approach could be to continue using the "followers" field, with the value of "zk_learners" in newer versions, but it'd be probably better to migrate to the new terminology used by Zookeeper.

Any of these changes would require more code after applying `schemaLeader`.

pfcoperez replied:
Thanks for the suggestion! In 97e5518a6d8409e2ad4c706e6c97e4d1d89d5a23 I added some logic to fill the "learners" field with "followers" (modern view of legacy ZK versions) or to fill the "followers" field with "learners" (back compatibility with old dashboards when using ZK>3.6).

@jsoriano The ITs only run ensembles with one ZK node, meaning that they are always "standalone" which implies that the fields "zk_followers" or "zk_learners" will not ever be present in the tested mntr outputs. I've tried to find, with no success, an unit test for the fields extraction. How do you unit test this code?

jsoriano reviewed (approved):
Approving, waiting on my last comment before merging, in case you want to do this now.

sayden commented on CHANGELOG.next.asciidoc:
```diff
@@ -111,6 +111,8 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...main[Check the HEAD dif
 - Extract correct index property in kibana.stats metricset {pull}29622[29622]
 - Fixed bug with `elasticsearch/cluster_stats` metricset not recording license expiration date correctly. {pull}29711[29711]
 - Enhance metricbeat on openshift documentation {pull}30054[30054]
+- Fixed GCP GKE Overview dashboard {pull}29913[29913]
```
This line should not be here

pfcoperez replied:
++ Nice catch! This is the result of a conflict resolution commit. Fixed in 93567d43245fa224a4226a5a8b28fcd6f56467bc

sayden commented on metricbeat/module/zookeeper/mntr/data.go:
```diff
@@ -86,8 +87,16 @@ func eventMapping(serverId string, response io.Reader, r mb.ReporterV2, logger *
 
 	// only exposed by the Leader
-	if _, ok := fullEvent["zk_followers"]; ok {
+	if hasFollowers, hasLearners := fullEvent["zk_followers"], fullEvent["zk_learners"]; hasFollowers != nil || hasLearners != nil {
```
This oneliner plus the block inside it is not very idiomatic in go and `hasFollowers` is not a boolean variable, it's an int so its name should be something like `numFollowers` in this case or something along those lines.

Please, write each case separatedly and define the situations where you are omitting some result with a comment. If you can omit some value, please write a comment explaining why. Example:

```go
if value != 7 {
    // we can safely ignore values under same because reason 1 and values over 7 because reason 2. Decide if it's worth to write a debug log (not info) in such case in case things get hairy in the future
}
doLogicWithSeven(value)
```
You have partially done it already, I just can't quickly why learners and followers are related in the source code to be in the same block.

sayden replied:
It's just a style request, not saying it's incorrect btw.

sayden reviewed (changes requested):
Requested one mandatory but small change and one style change if possible 🙂

sayden reviewed (approved):
All right! Thank you very much for the help!

pfcoperez commented:
@jsoriano @sayden Thanks a lot for your thorough review and insightful help.

In b8bee51b312860f177d1f37de496ce583044c8d9 I've added UTs, may I ask you to please take a look? If you find them OK i'll just merge.

jsoriano commented on metricbeat/module/zookeeper/mntr/mntr_test.go:
```diff
@@ -0,0 +1,616 @@
+		assert.Equal(t, float64(29), latencyFields["max"], versionMsg)
+		if sample.Version != "3.5.3" {
+			// Float precission is only provided from ZK 3.6
+			assert.Equal(t, float64(0.5714), latencyFields["avg"], versionMsg)
```
You could add a map of fields to compare in each `TestCase`, so you don't need conditional tests here.

It would be something like this in the definition:
```
	type TestCase struct {
		Version    string
		MntrSample string
		ExpectedValues map[string]interface{}
	}

	mntrSamples := []TestCase{
		{"3.5.3", mntrTestInputZooKeeper35, map[string]interface{
			"learners": int64(1),
			"latency.min": float64(0),
		}},
		{"3.7.0", mntrTestInputZooKeeper37, map[string]interface{
			"learners": int64(1),
			"latency.min": float64(0),
			"latency.avg": float64(0.5714),
		}},},
	}
```
And a loop here checking the fields.
```
for k, v := range sample.ExpectedValues {
        found, _ := event.MetricSetFields.Get(k)
	assert.Equal(t, v, found)
}
```

pfcoperez replied:
Thanks @jsoriano , I thought about this but I wanted to keep it simple. For example: Comparing flattened maps for `latency` to nested maps would require some extra effort.

I am giving it a try anyway.

pfcoperez replied:
Ahh nice! I had however already implemented a nested reports assertion function in 77212ad84c4dd198fe91de4c2f16aa94a5c90335

I'd like to keep it because I think it is easier to read nested maps than flattened ones (DRY :sweat_smile: ). With it, they look like:

```go
{
			"3.7.0",
			mntrTestInputZooKeeper37,
			common.MapStr{
				"learners":  int64(1),
				"followers": int64(1),
				"latency": common.MapStr{
					"max": float64(8),
					"avg": float64(0.5714),
					"min": float64(0),
				},
			},
		},
```

And the nested assertions checker looks like:

```go
func assertExpectations(t *testing.T, expectations common.MapStr, report common.MapStr, message ...string) {
	for key, expectation := range expectations {
		assert.Contains(t, report, key, message)
		switch expectation := expectation.(type) {
		case common.MapStr:
			nestedReport, _ := report.GetValue(key)
			assert.IsType(t, nestedReport, report, message)
			assertExpectations(t, expectation, nestedReport.(common.MapStr), message...)
		default:
			reportValue, _ := report.GetValue(key)
			assert.Equal(t, expectation, reportValue, message)
		}
	}
}
```

jsoriano reviewed (commented):
Thanks for adding unit tests! I have added some suggestions there, and I will leave the final word to @sayden :slightly_smiling_face:

pfcoperez commented on metricbeat/module/zookeeper/mntr/mntr_test.go:
```diff
@@ -0,0 +1,616 @@
+
+	logger := logp.NewLogger("mntr_test")
+
+	for i, sample := range mntrSamples {
```
I love this! I was actually implementing this manually wanting to have something like:

```scala
describe(s"version [$version]") {
  // Tests
}
```

Applying!

pfcoperez commented on metricbeat/module/zookeeper/mntr/mntr_test.go:
```diff
@@ -0,0 +1,616 @@
+	mbtest "github.com/elastic/beats/v7/metricbeat/mb/testing"
+)
+
+var mntrTestInputZooKeeper35 = `zk_version	3.5.3-beta-8ce24f9e675cbefffb8f21a47e06b42864475a60, built on 04/03/2017 16:19 GMT
```
I was wondering how to do this, definitively applying this suggestion!

jsoriano reviewed (commented):
LGTM

jsoriano commented:
/test

sayden reviewed (approved):
Thanks for all the effort!
